### ISSUE TITLE
WebSocket client occasionally not receiving first server frame

### ISSUE THREAD

**Issue by ryanpbrewster** [faye/faye-websocket-node#73]
Apologies in advance for the vagueness of this issue.

I have observed different client-side behavior between this package and the [`ws`](https://www.npmjs.com/package/ws) package. In particular, this package seems to sometimes drop or ignore messages sent very shortly after the websocket handshake.

Here is a minimal repro: https://gist.github.com/ryanpbrewster/6c28751425ec9c10c8e79ae2ded38255

Here is the output of the repro:
```
> node ws.js ws
connect: 12.882ms ok
connect: 2.403ms ok
connect: 2.038ms ok
connect: 0.979ms ok
connect: 1.908ms ok
connect: 2.008ms ok
connect: 2.256ms ok
connect: 3.504ms ok
connect: 2.916ms ok
connect: 2.741ms ok
connect: 2.617ms ok



> node ws.js faye
connect: 3.512ms ok
connect: 4.956ms ok
connect: 4.756ms ok
connect: 4.125ms ok
connect: 3.536ms ok
connect: 2.863ms ok
connect: 3.763ms ok
connect: 3.523ms ok
connect: 3.749ms ok
connect: 500.985ms timeout
connect: 6.456ms ok
connect: 5.415ms ok
connect: 5.455ms ok
connect: 4.845ms ok
```

In short, this example is:
  - start a simple websocket server in the background --- all it does it accept incoming connections, perform the websocket handshake, then immediately try to send a single frame containing `"hello"`
  - repeatedly create a client, negotiate the websocket handshake, and wait for the first frame from the server

Occasionally with `faye-websocket` I observe that the client never receives that initial frame. I cannot reproduce this behavior using a Node.js websocket server, so my initial thought was that it may be a bug in the server implementation. However, the `ws` module does not seem to have this issue.

Any insight into why the client sometimes does not receive these frames?

**Comment by jcoglan** [comment]
Just to clarify, is this a demonstration of a performance difference between `faye-websocket` and `ws`, or is it demonstrating a bug? I understand that the example demonstrates faye-websocket taking longer to receive the first message and close, thus resolving the promise you're measuring, but I'm not sure it shows any messages being dropped.

**Comment by ryanpbrewster** [comment]
I believe it is a bug. `faye-websocket` is timing out because it never receives the first message.

**Comment by jcoglan** [comment]
It could be because the server is pushing data at a rate faster than faye-websocket is processing it, so it doesn't emit a message event before the timeout you've set. I'm not saying that's not a problem, but we should try to characterise this bug in terms of whether it loses data or not. Can you construct an example that demonstrates the client dropping messages from the server? e.g. having the server sending a sequence of numbers and the client failing to print all of them?

**Comment by ryanpbrewster** [comment]
Here is an example run with a 15s timeout
```
attempt-72: 4.039ms ok
attempt-73: 3.612ms ok
attempt-74: 3.059ms ok
attempt-75: 3.312ms ok
attempt-76: 2.252ms ok
attempt-77: 3.412ms ok
attempt-78: 3.751ms ok
attempt-79: 3.265ms ok
attempt-80: 2.766ms ok
attempt-81: 15000.988ms timeout
attempt-82: 6.057ms ok
attempt-83: 4.643ms ok
attempt-84: 4.224ms ok
attempt-85: 4.096ms ok
attempt-86: 4.125ms ok
```

15s implies to me that the issue is actually a dropped/ignored message, not client lag or a performance issue.

**Comment by ryanpbrewster** [comment]
To clarify: this is only on the initial message. I have no seen any issues with dropped messages once the client receives the first one.

If I modify the server to send two messages, one immediately and one 500ms later, the `faye-websocket` client reliably receives the second message (but not always the first).

**Comment by jcoglan** [comment]
You're right, I misinterpreted your original demo. I've investigated and there are two causes here:

- The server sends `transfer-encoding: chunked` in its WebSocket handshake response, which is incorrect -- the body is not in chunked encoding
- The HTTP parser we use introduced a bug in version 0.4.11 that means that if the first message arrives in the same chunk of TCP input as the headers, then we lose that first message as the parser consumes it. I've opened an issue for that here: https://github.com/creationix/http-parser-js/issues/63

**Comment by jcoglan** [comment]
You can work around this issue by installing `http-parser-js@0.4.10` alongside `faye-websocket` so that you get a working version of the parser.

**Comment by jcoglan** [comment]
I've also just published version 0.7.2 of websocket-driver, which pins the http-parser-js dependency so that we're not affected by this bug.

**Comment by ryanpbrewster** [comment]
Issue resolved as of `v0.11.3`, nice analysis and thanks for the quick fix!
